### ISSUE TITLE
Can't find any good documentation describing "Huber scale"

### ISSUE THREAD

**Issue by akawana** [bmaltais/kohya_ss#3494]
If you have any documentation describing this parameter, or if you understand the math behind this setting well, please share your knowledge.

**Comment by AbstractEyes** [comment]
I had Claude Opus 4.6 write this one up for you.

<html>
<body>
<html><head></head><body><h1><code>huber_scale</code> — What it does and how to use it</h1>
<h2>Background</h2>
<p><code>huber_scale</code> was introduced in <a href="https://github.com/kohya-ss/sd-scripts/pull/1808">PR #1808</a> by @recris to support <strong>pseudo Huber loss</strong> for Flux.1 and SD3 training. The original Scheduled Huber Loss implementation (<a href="https://github.com/kohya-ss/sd-scripts/pull/1228">PR #1228</a>) used a fixed internal threshold that worked poorly with Flux/SD3's different latent error profiles.</p>
<h2>The Math</h2>
<p>The pseudo Huber loss function is:</p>
<pre><code>L(a) = δ² * (sqrt(1 + (a/δ)²) - 1)
</code></pre>
<p>Where <code>a</code> is the per-element prediction error (noise_pred - target) and <code>δ</code> (delta) is the threshold derived from <code>huber_scale</code>.</p>
<p><strong><code>huber_scale</code> controls δ — the transition point between L1-like and L2-like behavior:</strong></p>
<ul>
<li>When <code>|a| &lt;&lt; δ</code>: the loss behaves like <strong>L2 (MSE)</strong> — quadratic, sensitive to small errors, good for fine details</li>
<li>When <code>|a| &gt;&gt; δ</code>: the loss behaves like <strong>L1 (MAE)</strong> — linear, robust to outliers/corrupted data</li>
</ul>
<p>So <code>huber_scale</code> sets <em>where that crossover happens</em> in terms of the latent error magnitude.</p>
<h2>Practical Effect</h2>

huber_scale | Effect
-- | --
Lower (e.g. 1.0) | More aggressive dampening of large errors → more L1-like → more outlier-robust but slower learning, requires higher LR
Higher (e.g. 3.0+) | Less dampening → more L2-like → closer to standard MSE behavior


<p>Because the loss dampens large gradient contributions, <strong>you typically need to increase your learning rate by 20-50%</strong> when using Huber loss compared to standard L2.</p>
<h2>Relationship to Other Parameters</h2>
<ul>
<li><strong><code>loss_type</code></strong>: Must be set to <code>huber</code> or <code>smooth_l1</code> for any of this to take effect. Default is <code>l2</code> (standard MSE).</li>
<li><strong><code>huber_c</code></strong>: Controls the <strong>decay rate</strong> of the scheduled threshold over timesteps (when using <code>huber_schedule=exponential</code>). This is <em>not</em> the same as <code>huber_scale</code>. <code>huber_c</code> was the original parameter from PR #1228; <code>huber_scale</code> was added later because <code>huber_c</code> alone couldn't control the base threshold.</li>
<li><strong><code>huber_schedule</code></strong>: How the threshold varies with timestep — <code>snr</code>, <code>exponential</code>, or <code>constant</code>. Default is <code>snr</code>. The idea is to be more L1-like (outlier-robust) at high-noise timesteps and more L2-like (detail-preserving) at low-noise timesteps.</li>
</ul>
<h2>Recommended Starting Points</h2>
<p>Per the PR author (@recris) and community testing:</p>
<p><strong>Flux.1 LoRA:</strong></p>
<pre><code>--loss_type huber --huber_scale 2.5 --huber_c 1.0 --huber_schedule exponential
</code></pre>
<p>Earlier recommendation was <code>huber_scale=1.8, huber_c=0.25</code>, but later testing found <code>huber_scale=2.5, huber_c=1.0</code> (more aggressive exponential decay) reduced artifacts when scaling LoRA strength past 100% or stacking multiple LoRAs.</p>
<p><strong>SDXL LoRA:</strong>
The original defaults from PR #1228 (<code>huber_c=0.1</code>, <code>huber_schedule=snr</code>) still apply. <code>huber_scale</code> was not part of the original SDXL implementation and may not be needed there.</p>
<h2>Key Observations from Community Testing</h2>
<ul>
<li>Huber loss tends to produce <strong>slower but more consistent</strong> learning — the training concept gets learned more reliably over longer runs</li>
<li>It helps most with <strong>outlier resilience</strong> (corrupted/low-quality images in dataset), not necessarily with fine detail reproduction</li>
<li>You can switch from L2 to Huber mid-training and see immediate quality improvement — you don't have to start with it</li>
<li>If learning is too slow, increase <code>huber_scale</code> (or increase LR)</li>
</ul>
<h2>References</h2>
<ul>
<li><a href="https://github.com/kohya-ss/sd-scripts/pull/1808">PR #1808</a> — Pseudo Huber implementation for Flux/SD3 (where <code>huber_scale</code> was introduced)</li>
<li><a href="https://github.com/kohya-ss/sd-scripts/pull/1228">PR #1228</a> — Original Scheduled Huber Loss implementation</li>
<li><a href="https://github.com/kohya-ss/sd-scripts/discussions/294">Discussion #294</a> — "Everything you know about loss is a LIE!" (foundational discussion)</li>
</ul></body></html>
</body>
</html># `huber_scale` — What it does and how to use it

## Background

`huber_scale` was introduced in [[PR #1808](https://github.com/kohya-ss/sd-scripts/pull/1808)](https://github.com/kohya-ss/sd-scripts/pull/1808) by @recris to support **pseudo Huber loss** for Flux.1 and SD3 training. The original Scheduled Huber Loss implementation ([[PR #1228](https://github.com/kohya-ss/sd-scripts/pull/1228)](https://github.com/kohya-ss/sd-scripts/pull/1228)) used a fixed internal threshold that worked poorly with Flux/SD3's different latent error profiles.

## The Math

The pseudo Huber loss function is:

```
L(a) = δ² * (sqrt(1 + (a/δ)²) - 1)
```

Where `a` is the per-element prediction error (noise_pred - target) and `δ` (delta) is the threshold derived from `huber_scale`.

**`huber_scale` controls δ — the transition point between L1-like and L2-like behavior:**

- When `|a| << δ`: the loss behaves like **L2 (MSE)** — quadratic, sensitive to small errors, good for fine details
- When `|a| >> δ`: the loss behaves like **L1 (MAE)** — linear, robust to outliers/corrupted data

So `huber_scale` sets *where that crossover happens* in terms of the latent error magnitude.

## Practical Effect

| huber_scale | Effect |
|---|---|
| **Lower** (e.g. 1.0) | More aggressive dampening of large errors → more L1-like → more outlier-robust but slower learning, requires higher LR |
| **Higher** (e.g. 3.0+) | Less dampening → more L2-like → closer to standard MSE behavior |

Because the loss dampens large gradient contributions, **you typically need to increase your learning rate by 20-50%** when using Huber loss compared to standard L2.

## Relationship to Other Parameters

- **`loss_type`**: Must be set to `huber` or `smooth_l1` for any of this to take effect. Default is `l2` (standard MSE).
- **`huber_c`**: Controls the **decay rate** of the scheduled threshold over timesteps (when using `huber_schedule=exponential`). This is *not* the same as `huber_scale`. `huber_c` was the original parameter from PR #1228; `huber_scale` was added later because `huber_c` alone couldn't control the base threshold.
- **`huber_schedule`**: How the threshold varies with timestep — `snr`, `exponential`, or `constant`. Default is `snr`. The idea is to be more L1-like (outlier-robust) at high-noise timesteps and more L2-like (detail-preserving) at low-noise timesteps.

## Recommended Starting Points

Per the PR author (@recris) and community testing:

**Flux.1 LoRA:**
```
--loss_type huber --huber_scale 2.5 --huber_c 1.0 --huber_schedule exponential
```

Earlier recommendation was `huber_scale=1.8, huber_c=0.25`, but later testing found `huber_scale=2.5, huber_c=1.0` (more aggressive exponential decay) reduced artifacts when scaling LoRA strength past 100% or stacking multiple LoRAs.

**SDXL LoRA:**
The original defaults from PR #1228 (`huber_c=0.1`, `huber_schedule=snr`) still apply. `huber_scale` was not part of the original SDXL implementation and may not be needed there.

## Key Observations from Community Testing

- Huber loss tends to produce **slower but more consistent** learning — the training concept gets learned more reliably over longer runs
- It helps most with **outlier resilience** (corrupted/low-quality images in dataset), not necessarily with fine detail reproduction
- You can switch from L2 to Huber mid-training and see immediate quality improvement — you don't have to start with it
- If learning is too slow, increase `huber_scale` (or increase LR)

## References

- [[PR #1808](https://github.com/kohya-ss/sd-scripts/pull/1808)](https://github.com/kohya-ss/sd-scripts/pull/1808) — Pseudo Huber implementation for Flux/SD3 (where `huber_scale` was introduced)
- [[PR #1228](https://github.com/kohya-ss/sd-scripts/pull/1228)](https://github.com/kohya-ss/sd-scripts/pull/1228) — Original Scheduled Huber Loss implementation
- [[Discussion #294](https://github.com/kohya-ss/sd-scripts/discussions/294)](https://github.com/kohya-ss/sd-scripts/discussions/294) — "Everything you know about loss is a LIE!" (foundational discussion)

**Comment by akawana** [comment]
Thank you, you’re the right kind of person :)! I really appreciate it when answers are given in proper documentation form.

This actually clarified quite a lot, because I had read about huber loss in general, but couldn’t figure out the parameters. And in some cases, when training a LoRA, hyper loss is extremely useful.

Thank you!

**Comment by akawana** [comment]
The only thing I can add from my own experience is that for **SDXL** specifically, Huber and Smooth L1 work great when the Huber schedule is set to **exponential**.

**Comment by bmaltais** [comment]
Since an answer has been provided, I will close this issue.
